src/vm/database/key_value_wrapper.rs

```diff
@@ -0,0 +1,146 @@
+    // the underlying key-value storage.
+    store: Box<KeyValueStorage + 'a>,
+    // lookup_map is a history of edits for a given key.
+    //   in order of least-recent to most-recent at the tail.
+    //   this allows ~ O(1) lookups, and ~ O(1) commits, roll-backs (amortized by # of PUTs).
+    lookup_map: HashMap<String, VecDeque<String>>,
+    // stack keeps track of the most recent rollback context, which tells us which
+    //   edits were performed by which context. at the moment, each context's edit history
+    //   is a separate Vec which must be drained into the parent on commits, meaning that
+    //   the amortized cost of committing a value isn't O(1), but actually O(k) where k is
+    //   stack depth.
+    //  TODO: The solution to this is to just have a _single_ edit stack, and merely store indexes
+    //   to indicate a given contexts "start depth".
+    stack: VecDeque<RollbackContext>
+}
+
+impl <'a> RollbackWrapper <'a> {
+    pub fn new(store: Box<KeyValueStorage + 'a>) -> RollbackWrapper {
+        RollbackWrapper {
+            store: store,
+            lookup_map: HashMap::new(),
+            stack: VecDeque::new()
+        }
+    }
+
+    pub fn nest(&mut self) {
+        self.stack.push_back(RollbackContext { edits: Vec::new() });
+    }
+
+    // Rollback the child's edits.
+    //   this clears all edits from the child's edit queue,
+    //     and removes any of those edits from the lookup map.
+    pub fn rollback(&mut self) {
+        let mut last_item = self.stack.pop_back()
+            .expect("ERROR: Clarity VM attempted to commit past the stack.");
+
+        last_item.edits.reverse();
+
+        for (key, value) in last_item.edits.drain(..) {
+                let remove_edit_deque = {
+                    let key_edit_history = self.lookup_map.get_mut(&key)
+                        .expect("ERROR: Clarity VM had edit log entry, but not lookup_map entry");
+                    let popped_value = key_edit_history.pop_back();
+                    assert_eq!(popped_value.as_ref(), Some(&value));
+                    key_edit_history.len() == 0
+                };
+                if remove_edit_deque {
+                    self.lookup_map.remove(&key);
+                }
+        }
+    }
+
+    pub fn commit(&mut self) {
+        let mut last_item = self.stack.pop_back()
+            .expect("ERROR: Clarity VM attempted to commit past the stack.");
+
+        if self.stack.len() == 0 {
+            // committing to the backing store
+            for (key, value) in last_item.edits.drain(..) {
+                let remove_edit_deque = {
+                    let key_edit_history = self.lookup_map.get_mut(&key)
+                        .expect("ERROR: Clarity VM had edit log entry, but not lookup_map entry");
+                    let popped_value = key_edit_history.pop_front();
+                    assert_eq!(popped_value.as_ref(), Some(&value));
+                    self.store.put(&key, &value);
```

Comment:
The reason the MARF has an `insert_batch()` method is to make this kind of operation much faster.  Each time you call `insert()` on the MARF, you're re-calculating the Merkle skip-list root.  `insert_batch()` only does this once per batch, instead of once per key/value pair.

Reply:
Ahh, yes, this is a perfect use case for `insert_batch`, I'll update.